GcExtensions.md

```diff
@@ -0,0 +1,353 @@
+  layouts.* Not all objects -- especially those from preexisting
+  libraries -- fit Julia's type system, for example, specialized
+  container types written in C/C++. Such objects can comprise multiple
+  memory blocks that require a custom marking mechanism and may also
+  require finalizers written in C.
+* *Providing additional roots to the GC.* Currently, to have additional
+  roots, they must be stored in a location that is visible to Julia.
+  This can be expensive if such roots are updated frequently or are
+  contained in data structures that would have to be laboriously
+  translated into a format usable by Julia. Instead, we want to allow
+  for roots to be discoverable at the beginning of a garbage collection.
+* *Conservative scanning of stack frames and objects.* Currently,
+  scanning does have to be precise. If we desire to use the GC for
+  foreign code that requires conservative scanning (especially for
+  foreign stack frames), then it is necessary to have functionality
+  that determines whether a machine word is a pointer to an object,
+  including to its interior.
+
+To demonstrate the applicability and viability of these mechanisms, we
+have fully integrated Julia with the GAP computer algebra system, to the
+point that GAP's regular garbage collector is completely replaced with
+Julia's and that the lifetime of all GAP objects is entirely managed by
+Julia. We also implemented a self-contained test program using these
+mechanisms, and integrated it into the Julia test suite.
+
+The proposed implementation should not incur measurable overhead for
+Julia itself, as it only exposes additional functionality that is unused
+by Julia code, plus functionality hooks that are designed to only incur
+a few clock cycles of overhead per garbage collection. More specific
+discussion of overhead can be found accompanying the descriptions of
+these hooks.
+
+An implementation of this proposal can be found on GitHub under
+https://github.com/rbehrends/julia (branch `rb/gc-extensions`). See
+the example in the `test/gcext` subdirectory for an example of using
+this API.
+
+An implementation of GAP that uses the Julia GC in lieu of its native
+GC can likewise be found on GitHub at https://github.com/rbehrends/gap
+(branch `alt-gc`). This version of GAP can be built with:
+
+    ./autogen.sh
+    ./configure --with-gc=julia --with-julia=/path/to/julia/usr
+    make
+
+## Callbacks
+
+In order to allow foreign code to have access to necessary functionality
+in the garbage collector, we allow foreign code to register callbacks for
+certain GC events. We provide for six types of callbacks:
+
+* `pre_gc` (beginning of garbage collection).
+* `post_gc` (end of garbage collection).
+* `root_scanner` (before scanning GC roots).
+* `task_scanner` (when scanning Julia tasks).
+* `notify_external_alloc` (when an external object is allocated).
+* `notify_external_free` (when an external object is freed).
```

Comment:
Everyone of these can be called from any thread. Everyone other than `pre_gc` and `post_gc` can be called concurrently.

As for the current implementation, `post_gc` may not happen before the next `pre_gc`, which may or may not be what you want.

Reply:
I'm aware of that (well, not the part that `post_gc` may happen before the next `pre_gc`). But yes, general implementations would have to be thread-safe. I will amend the text to make a note of that.

Reply:
Finallizers can run arbitrary julia code, make arbitrary allocations and must be run with the GC enabled. That's why multiple GC can happen and finish before hitting the `post_gc` callback, **even with one thread**. If that's not what you want, you may need to use another list instead of the normal finalizer run list.

Reply:
That's actually totally fine by us. Our finalizers right now are plain C functions which take great care to not trigger another GC in any way.

Reply:
Other finalizers could still trigger GC you the ordering will still not be guaranteed. If that's OK, just documening the behavior should be fine.